controllers/topology/desired_state.go

```diff
@@ -170,6 +182,106 @@ func computeCluster(current *clusterTopologyState, infrastructureCluster, contro
 	return cluster
 }
 
+// computeMachineDeployment computes the desired state for a MachineDeploymentTopology.
+// The generated machineDeployment object is calculated using the values from the machineDeploymentTopology and
+// the machineDeployment class.
+func computeMachineDeployment(class *clusterTopologyClass, current *clusterTopologyState, topology clusterv1.MachineDeploymentTopology) (machineDeploymentTopologyState, error) {
+	desiredMachineDeployment := machineDeploymentTopologyState{}
+
+	className := topology.Class
+	machineDeploymentClass, err := getMachineDeploymentClass(class, className)
+	if err != nil {
+		return desiredMachineDeployment, err
+	}
+
+	machineDeploymentTopologyClass, ok := class.machineDeployments[className]
+	if !ok {
+		return desiredMachineDeployment, errors.Errorf("class name %s not found in clusterClass %s", className, class.clusterClass.Name)
+	}
+
+	currentMachineDeployment := current.machineDeployments[topology.Name]
+	var currentBootstrapTemplate *corev1.ObjectReference
+	if currentMachineDeployment.bootstrapTemplate != nil {
```

Comment:
Q: should we check if the current machine deployment is nil (there is no current machine deployment)?
(same for the infrastructureRef)

Reply:
This is covered in the first test when no previous machineDeployment exists. It returns an empty instance of `machineDeploymentTopologyState` with all the 3 values set to nil.

Reply:
WDYT about also checking for `currentMachineDeployment.object != nil`? I know it shouldn't happen as `bootstrapTemplate` and `object` are always set together, but just to make it a bit more safe against panic's.